### DISCUSSION TITLE
Project future, maintainability concerns and suggestions on how to improve the project

### DISCUSSION THREAD

**Discussion by frica** [TomSchimansky/CustomTkinter#2343] [General]
Hello everyone,

_Disclaimer: I have a lot of respect for open-source maintainers and I'm grateful to @TomSchimansky for this toolkit. There is certainly some context I miss and I apologize if it was discussed before._

I came across CTk very recently. I've been looking in the last months at many UI frameworks for Python. Tested a few, looked at their repo and also how active they are. Some frameworks look good, some are easier than others, most modern solutions are "web-based" (= involve some kind of web-server for the rendering) which is overkill in some case for desktop applications.

I never liked Tk. The default UI output has been always bad, whether it was 20 years ago or today. But **CTk fixes that rendering issue** and make it more rewarding to overcome the Tk layouting logic. The documentation is also pretty good. So thanks to the people involved for all the efforts!

But:
- There are bugs though. 330 as I write this. 
- 47 PR are open today.
- There have been 5 PRs from external contributors merged to main the last 12 months.

So I see only one maintainer. I see PRs waiting for a while. I see active and knowledgeable people in the issues discussions but no one has the authorization to triage the incoming issues, nor merge the valuable PRs. I see no plan for the future.

I don't know how it's organized internally between the maintainer and the long time supporters/technical experts of CTk. But I think it's possible to improve the repository and project state with a few steps:
- add a file to explain how to help (see example [here](https://github.com/zauberzeug/nicegui/blob/main/CONTRIBUTING.md))
- set a better template for the issues to avoid low-quality bug reports
- give rights to a few people on Issues/PR/Discussions. This would allow for instance to close faster obvious irrelevant defects/PR (like https://github.com/TomSchimansky/CustomTkinter/pull/2287 or https://github.com/TomSchimansky/CustomTkinter/issues/2311), move discussion posts from one category to another, tag issues as bug, enhancement etc. This would help the maintainer to have a cleaner state when he has time to look at the repository
- Let contributors update documentation by providing publicly the documentation files  
- Allow 1 or 2 key contributors (chosen by the maintainer of course) to merge PRs (to be decided which ones)

Tell me what you think about these ideas and let's openly discuss about the topic in a polite and constructive manner.

Regards,

Fabien

**Comment by JanPanthera** [comment]
Maybe we someday create our own, since Tom seems not active anymore

**Reply by frica** [reply]
That's the last option, and we should really to avoid it, according to me.
Tom committed 2 months ago so I'll give him some time to review the discussion.

**Reply by JanPanthera** [reply]
I mean we could write our own using opengl or even vulkan, 😹
This would be way faster then tk xD

**Reply by frica** [reply]
That's an ambitious idea, way above what I can offer as help 😄

**Reply by JanPanthera** [reply]
Me too, but it would be fun. Some C++ or doing it in python, i am really mad with this slow tk an this overblown qt lol

**Reply by Akascape** [reply]
@JanPanthera I also found this [dearpygui](https://github.com/hoffstadt/DearPyGui) library based on c++, which is actually good in performance. It's perfect for building feature rich python projects.

**Reply by JanPanthera** [reply]
From the name it hears like rip off from dearimgui xD, anyway i need to complete my current project until i can start have fun with c++ or opengl/vulkan in python 😹
I coul be done with my tool but i strangely have to code my own widgets.... TreeView... hust hust

**Reply by TomatoScientist** [reply]
> @JanPanthera I also found this [dearpygui](https://github.com/hoffstadt/DearPyGui) library based on c++, which is actually good in performance. It's perfect for building feature rich python projects.

My first GUI library in Python was "dearpygui", but I'm confused by the excessive fanaticism around the context manager. I think CTk structure is more understandable and logical :))

**Comment by Akascape** [comment]
@TomSchimansky If you wish, I can manage the issue and discussions. 
I regularly check the discussions and issues tab, so I will try to close unnecessary/duplicate issues, convert some of them to discussions, and closing fixed issues.

**Reply by TomSchimansky** [reply]
I need to move the repository to an organisation in order to assign people with moderation rights only but i will think about it 👍

**Reply by TomSchimansky** [reply]
If you are still interested being a collaborator as well to improve the project further, merge pull requests, answer issues, fix bugs, you can write me at: [contact@customtkinter.tomschimansky.com](mailto:contact@customtkinter.tomschimansky.com)

**Reply by Akascape** [reply]
@TomSchimansky Yes, I’m ready to continue my contribution for CustomTkinter. I have sent the email. I believe it is still one of the easiest modern UI libraries for Python.

**Reply by Akascape** [reply]
@TomSchimansky And I think we need a developer group first, in any platform like discord.

**Reply by Akascape** [reply]
@TomSchimansky Please respond, I have already sent the request.

**Reply by dipeshSam** [reply]
@TomSchimansky 

I’m genuinely interested in collaborating further and would be happy to contribute more time to:

fixing bugs,
addressing open issues,
improving documentation, and
reviewing or submitting pull requests where helpful.
Thank you for your work on the project, and I look forward to contributing again.

Best regards

**Comment by TomSchimansky** [comment]
Hi @frica, I briefly wanted to let you know that I noticed your post and I am totally aware of the situation and the problems you mentioned. I am finishing my Bachelor thesis at the moment and I dont have any time left to work on this project at the moment. However this will change in 5 weeks and then I will start to clean up the issues and address the most important bugs. I never could bring myself to the point of assigning collaborators because I thought, and still think, that this project is not suited very well for collaborative development. There are so many workarounds and hacks used to make this library work which I dont understand myself after some time. And if one merges a pull request without being aware of these or without proper testing it could be that functionality breaks but only on certain operating systems and so on. The makeshift nature of this project is also what kept me from working further on the project because I saw no future for it because of all the workarounds and the disappointing performance. But as so many people are still using the library I might get motivated enough after my thesis to start addressing the most important bugs and feature requests. I will also think about assigning collaborators, at least for moderation. Only a little patience is needed until then. 🙏

**Reply by frica** [reply]
Hey Tom, 

Thanks for taking the time to respond. As stated in my original post I didn't know much of the context and I'm glad you could clarify the state of the project. My suggestions were meant to provide some help and guidance, and I'm not pretending I could do better at all :)

I also understand your situation, no worries. My point of view is that your toolkit seems to have quite some users, interested in the nice look & feel and not really in performance. So it's still relevant! 

I will continue to follow your project and I wish you all the best for the bachelor thesis.

**Reply by ElSaico** [reply]
That's a very elucidating response.

I stumbled upon CTk as an option for the aging UI of [Elite: Dangerous Market Connector](https://github.com/EDCD/EDMarketConnector), an application which has also over time fostered an active ecosystem around it - and, very importantly, changed maintainers more than once.
This could allow us to throw away hundreds of lines of cruft on both its very hacky dark mode implementation and the platform-specific stuff that CTk already handles, while remaining backwards compatible with existing plugins (which make any non-Tkinter solution a very last choice). Alas, as a Portuguese speaker I was immediately stopped by the canvas locale bug (#571), and proceeded with a pure Ttk solution.

However, progress has slowed to a crawl because of Tk shenanigans, so it might make sense to wait for a new release and try it out instead.
Being on the final stretch of a thesis is a more than understandable reason to not have time left for anything else, so let me just address a couple points:

> I never could bring myself to the point of assigning collaborators because I thought, and still think, that this project is not suited very well for collaborative development. There are so many workarounds and hacks used to make this library work which I dont understand myself after some time.

The source code for any long-running piece of software can become a wild ride and a half, yet life goes on regardless.
EDMC is on its third (fourth?) maintainer - his predecessor remains around, but not the original creator, so _HIC SVNT DRACONES_ on some places. And I dove straight into a UI overhaul of it, for crying out loud! 😆

> And if one merges a pull request without being aware of these or without proper testing it could be that functionality breaks but only on certain operating systems and so on.

A fair concern, and one that EDMC constantly faces. One way to alleviate it could be by strengthening and automating its test suite, so that it runs on Actions - yes, it'd take significant work _(and does it allow a practical way to do Selenium-esque functional testing for desktop apps?)_, but it pays off on stability and peace of mind.

> The makeshift nature of this project is also what kept me from working further on the project because I saw no future for it because of all the workarounds and the disappointing performance.

Those are part and parcel for any Tkinter UI of a meaningful size, so why not concentrate improvements on a focal point instead of everyone needing to figure out a solution with a very high chance to turn out worse?

> But as so many people are still using the library I might get motivated enough after my thesis to start addressing the most important bugs and feature requests. I will also think about assigning collaborators, at least for moderation. Only a little patience is needed until then. 🙏

Given the timeframe, my guess is that you're probably still decompressing from the thesis effort. Take your time, the community you fostered (with several libraries out there proudly announcing CTk support, a good indicator of success IMHO) will be here waiting.

**Reply by TomatoScientist** [reply]
@TomSchimansky I really appreciate you for your library, because as a designer (in some sense) I was searching for GUI-tools, which allows to get a modern and customizable appearance. I believe that GUI Python projects are extremely underrated, so on my Python-lessons I show students how they can create pretty complex applications with CTk. Please, do not forget us 😄

**Reply by TomSchimansky** [reply]
Thanks to @FedericoSpada there will be some progress for this project again in the coming days and weeks!

**Comment by Snxwman** [comment]
Has any decision been made regarding whether this project will continue?

**Reply by O2H-Oscar** [reply]
> ¿Se ha tomado alguna decisión sobre si este proyecto continuará?

Hola.
Si, yo tengo la misma duda, creo que es fácil de implementar en proyectos pequeños y medianos, lo que le da mucho valor, pero si el proyecto está muerto, dedicarle tiempo de estudio no tiene mucho sentido.

En caso de no seguir, gracias por la labor realizada hasta aquí.

**Comment by notready123** [comment]
Hello, 
Is customtkinter project dead ? 

@TomSchimansky any news ?

Best regards

**Reply by TomSchimansky** [reply]
Will start fixing some bugs in the next weeks

**Reply by dipeshSam** [reply]
@TomSchimansky We are waiting eagerly. 🥳

**Reply by FedericoSpada** [reply]
@TomSchimansky news?
Perhaps it's better to work in a way that allows others to assist you with the effort.

**Reply by TomSchimansky** [reply]
Thanks to @FedericoSpada there will be some progress for this project again in the coming days!

**Reply by TomSchimansky** [reply]
If there are people out there really into this project and willing to invest some time and be a collaborator as well to improve the project further, merge pull requests, answer issues, fix bugs, you can write me at: contact@customtkinter.tomschimansky.com

**Reply by dipeshSam** [reply]
@TomSchimansky 

I’m genuinely interested in collaborating further and would be happy to contribute more time to:
- fixing bugs,
- addressing open issues,
- improving documentation, and
- reviewing or submitting pull requests where helpful.

Thank you for your work on the project, and I look forward to contributing again.

Best regards

**Comment by FedericoSpada** [comment]
Hello everyone,
It is with deep regret that I inform you that this repository is most likely dead.
I volunteered last December to continue the development, but since the end of January, the new version has been under review. It seems that Tom doesn't have the desire either to continue with the project, nor to let it live without him. This is totally understandable, but also a little bit sad for those who relied on this library.
Since I still believe that this project has the potential to be the primary solution for developing the GUI of simple projects, I've decided to clone the repository and continue the development. I've called it Custom**2**kinter.
I've just uploaded v5.3.0 to [PyPI](https://pypi.org/project/custom2kinter/), and, in the develop branch on [GitHub](https://github.com/FedericoSpada/Custom2kinter), I've already added type hints to the whole project and reworked ThemeManager to allow custom theme_keys, but this is just the beginning of all the changes I have in mind.

I'm looking for volunteers who can help me recreate the documentation on Tom's Website elsewhere and write and run tests for the upcoming releases.
You can contact me via Discord (user name: ldragon135) or [LinkedIn](www.linkedin.com/in/federicospada13).

Stay tuned!

**Reply by Varos8** [reply]
Dear @FedericoSpada ,
Thank you for continuing this project. I am a beginner python user and wanted to learn some simple GUI python package in case I might need it in the future. I was dissapointed to see that this project which seemed to be the best canidate for my needs seemed to be abandoned even though it is widely documented and used. 

I am still learning Tkinter as a foundation, but I am wondering if you believe that learning existing CTkinter syntax with existing resources will still be viable in the future i.e. do you intend to keep the Interface Framework similar/ backwards compatible ?

Best wishes

**Reply by FedericoSpada** [reply]
Hi,
It won't be completely backward-compatible, but the general syntax and code structure will be the same. Basically, some attributes and method names will be different, so existing projects require a little work if you decide to upgrade them to newer versions.
So, you can still learn CustomTkinter and apply most of the same knowledge to Custom2kinter. Moreover, the first released version (the one currently on PyPI) is backward-compatible and [mostly fixes some problems](https://github.com/FedericoSpada/Custom2kinter/blob/master/CHANGELOG.md) with the last CustomTkinter version.

**Comment by dipeshSam** [comment]
@FedericoSpada 
Agreed. Please keep [Github](https://github.com/FedericoSpada/Custom2kinter) and [PyPI](https://pypi.org/project/custom2kinter/) continue.

Best regards.
